src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Authentication.MSALWrapper.AuthFlow
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Microsoft.Extensions.Logging;
+
+    /// <summary>
+    /// The auth flows class.
+    /// </summary>
+    public class AuthFlowExecutor : IAuthFlow
+    {
+        private readonly IEnumerable<IAuthFlow> authflows;
+        private readonly ILogger logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AuthFlowExecutor"/> class.
+        /// </summary>
+        /// <param name="logger">The logger.</param>
+        /// <param name="authFlows">The list of auth flows.</param>
+        public AuthFlowExecutor(ILogger logger, IEnumerable<IAuthFlow> authFlows)
+        {
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            this.authflows = authFlows ?? throw new ArgumentNullException(nameof(authFlows));
+        }
+
+        /// <summary>
+        /// Gets the auth flow result.
+        /// </summary>
+        /// <returns>The <see cref="Task"/>.</returns>
+        public async Task<AuthFlowResult> GetTokenAsync()
+        {
+            AuthFlowResult result = new AuthFlowResult(null, new List<Exception>());
+            foreach (var authFlow in this.authflows)
+            {
+                var attempt = await authFlow.GetTokenAsync();
+                if (attempt == null)
+                {
+                    result.Errors.Add(new Exception("This is a catastrophic failure. AuthFlow result is null!"));

```

Comment:
Is it common in C# to just use a plain `Exception` like that? I only ask because in my experience with other languages it's more common in a situation like this to create one's own exception class.

Reply:
If this error bubbles up to the user (which I think it does), will it actually be helpful to them? 🤔

Reply:
Also, is it really catastrophic if it's happening in a loop where we might recover?

Reply:
Maybe something more like
```csharp
result.Errors.Add(new Exception($"Failed to get an auth token from ${authFlow.Name}"))
```
or, in lieu of `authFlow.Name` the equivalent of `typeof` or something in C#?

Reply:
+1, It is less common to use `Exception` in this way, and I agree making our own or at least naming the AuthFlow would be nice. There are a few ways to do that, one of them would be adding a Name method to the IAuthFlow interface which I would support. Every AuthFlow must implement it's get token method, and a Name method. Open to other ideas though

Reply:
I have created a work item and added to shield bugs to create a custom class for this exception. I discussed this with Kyle and wanted to scope this feature work since there's tons of enhancements that we can do. 

It is catastrophic because we get a null AuthFlowResult which is highly unlikely since it's composed of TokenResult and error list. Having a null TokenResult is common but having a null error list is not, so something got to be really off if it did happen. I had unit tests to cover cases when AuthFlowResult itself returns null.

Reply:
Scoping that to another PR is totally reasonable, @shalinikhare27. Sounds good to me.

As far as the catastrophe goes, I was thinking that failing **all** auth flows is the real catastrophe and failing an individual one, while troubling, is more of an inconvenience.

@kyle-rader: I think we can get away with a `nameof` for now instead of a `.Name()` method on that interface. That can be implemented now with no real impact on anything else @shalinikhare27 has merged.

Reply:
Kyle pushed a change for this. :)